Encrypted keystore

ChristopherDedominici opened the pull request:
For a list of follow-up tasks related to user experience improvements (not security improvements), see the [parent issue here](https://github.com/NomicFoundation/hardhat/issues/5722)

## QUESTIONS:
- user inputs are handled via `requestSecretInput` from `hre.interruptions`, but outputs are not currently managed by `hre.interruptions`. I believe they should be. Should we address this in this PR, or should I add it to the follow-up tasks?

ChristopherDedominici commented on v-next/hardhat-errors/src/descriptors.ts:
```diff
@@ -706,25 +706,12 @@ Try using another mnemonic or deriving less keys.`,
     },
   },
   KEYSTORE: {
-    INVALID_KEYSTORE_FILE_FORMAT: {
```
Removed because it is no longer needed, as user input is now handled via the `requestSecretInput` hook.

alcuadrado replied:
I don't get this comment. Maybe it's in the wrong line?

ChristopherDedominici replied:
Yes, This comment is for the following lines, the ones where I removed `USERINTERRUPTION_NOT_IMPLEMENTED`.

`INVALID_KEYSTORE_FILE_FORMAT` has been removed because the validation is now performed via the HMAC and not via zod, so an encryption error will be thrown

alcuadrado replied:
Oh, I see. I would expect the `INVALID_READLINE_OUTPUT` to also be removed

alcuadrado replied:
as in: this plugin shouldn't need readline

ChristopherDedominici commented on v-next/hardhat-keystore/src/internal/hook-handlers/configuration-variables.ts:
```diff
@@ -39,7 +41,16 @@ export default async (): Promise<Partial<ConfigurationVariableHooks>> => {
 
-      return keystore.readValue(variable.name);
+      const password = await askPassword(
+        context.interruptions.requestSecretInput.bind(context.interruptions),
```
`bind` is needed because `requestSecretInput` is a method of `hre.interruptions`, and when passed as an argument, it loses its original this context. Since method calls rely on the object reference for this, passing it as a standalone function removes that association. To ensure it retains the correct context, we need to explicitly bind this using `.bind(hre.interruptions)`.

This approach will be used also later in other files

alcuadrado replied:
Won't this ask for the password every time?

I think we want to cache the masterkey in line 17

ChristopherDedominici replied:
Yes, good catch. I'll address it and add a test to be sure the password is only asked once

ChristopherDedominici replied:
Done [here](https://github.com/NomicFoundation/hardhat/pull/6232/commits/dbaae55973f019c0520f48e72c9fa72b8901cd7a)

ChristopherDedominici commented on v-next/hardhat-keystore/src/internal/keystores/encryption.ts:
```diff
@@ -0,0 +1,630 @@
+  password: string;
+  salt: Uint8Array;
+}): Uint8Array {
+  password = password.normalize(PASSWORD_NORMALIZATION_FORM);
```
I need a double check that the `normalization` is only required here

alcuadrado replied:
Yes, I think this is correct.

ChristopherDedominici commented on v-next/hardhat-keystore/src/internal/keystores/password.ts:
```diff
@@ -0,0 +1,58 @@
+  ) => Promise<string>,
+  consoleLog: (text: string) => void = console.log,
+): Promise<string> {
+  const PASSWORD_REGEX = /^.{8,}$/;
```
Currently, we only require a minimum of 8 characters. Do we want to introduce additional rules?

agostbiro replied:
This is in line with [OWASP recommendation.](https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html#implement-proper-password-strength-controls) If we want to improve on this, we could provide password strength feedback to the user or [check for compromised password](https://haveibeenpwned.com/API/v3#PwnedPasswords), but I think these are nice to haves only.

alcuadrado replied:
agreed to start with this

ChristopherDedominici commented on v-next/hardhat-keystore/src/internal/keystores/password.ts:
```diff
@@ -0,0 +1,58 @@
+): Promise<string> {
+  const PASSWORD_REGEX = /^.{8,}$/;
+
+  consoleLog(UserDisplayMessages.keystoreBannerMessage());
```
How it currently looks like:
![image](https://github.com/user-attachments/assets/c65716de-0b96-42ee-9c3e-4a264cabac77)

ChristopherDedominici commented on v-next/hardhat-keystore/src/internal/keystores/password.ts:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  let confirmPassword: string | undefined;
+  while (confirmPassword === undefined) {
```
This loop will continue indefinitely until the confirmation password matches the original password.
This check ensures that the user saves the correct password, preventing spelling mistakes.

ChristopherDedominici commented on v-next/hardhat-keystore/test/keystores/encryption.ts:
```diff
@@ -0,0 +1,992 @@
+  });
+});
+
+describe("password normalization", () => {
```
I added this additional test to validate that the normalization is working

alcuadrado commented on v-next/hardhat-keystore/src/internal/keystores/keystore.ts:
```diff
@@ -0,0 +1,69 @@
+    return this.#keystoreData;
+  }
+
+  public async listKeys(): Promise<string[]> {
```
This is expected

alcuadrado commented on v-next/hardhat-keystore/src/internal/keystores/encryption.ts:
```diff
@@ -0,0 +1,630 @@
+import { siv } from "@noble/ciphers/aes";
```
I'll get to that tomorrow

alcuadrado commented:
> user inputs are handled via requestSecretInput from hre.interruptions, but outputs are not currently managed by hre.interruptions. I believe they should be. Should we address this in this PR, or should I add it to the follow-up tasks?

Actually, I think they are fine now. As the output is printed from tasks.

alcuadrado commented:
Still reviewing this, but looks good.

I think something that we should do is rename `key` to `id`.

ChristopherDedominici commented:
> Still reviewing this, but looks good.
> 
> I think something that we should do is rename the secret's `key` to `id`.

@alcuadrado yes, I added this as a follow up task with another one too, see here: https://github.com/NomicFoundation/hardhat/issues/5722

I added them as follow-up tasks in case we want to close this PR ASAP. However, if there's time, I'll address them right away

agostbiro reviewed (commented):
Hey, great job on this PR! 

I have one concern regarding HMAC verification of the file on disk. 

I noticed in a [comment](https://github.com/NomicFoundation/hardhat/pull/6232/files#r1950952694) that we don't want to verify the HMAC when listing keys to avoid prompting the user for a password. I think this is ok if the sole purpose of listing the keys is to display them to the user and it comes with a warning that the authenticity of the file is not verified.

But if there is further logic depending on the presence of a key and if an attacker can control whether a key is present, that is dangerous. For example, if a fallback mechanism is introduced in the future, the attacker can trigger the fallback mechanism by removing a key from the file on disk.

So I'd suggest creating two incompatible key store types to be used in different tasks: a read only `UnverifiedKeyStore` and read/write `VerifiedKeyStore` where read operations are verified too. Then it's easy to catch in code review whether the specific task needs verification and if the proper warnings are present.

agostbiro commented on v-next/hardhat-keystore/src/internal/keystores/keystore.ts:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  public async listKeys(): Promise<string[]> {
+    return Object.keys(this.#keystoreData.secrets);
```
Linking review comment: https://github.com/NomicFoundation/hardhat/pull/6232#pullrequestreview-2611368134

agostbiro commented on v-next/hardhat-keystore/src/internal/keystores/keystore.ts:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  public async hasKey(key: string): Promise<boolean> {
+    return Object.keys(this.#keystoreData.secrets).includes(key);
```
Linking review comment: https://github.com/NomicFoundation/hardhat/pull/6232#pullrequestreview-2611368134

ChristopherDedominici replied:
See here: https://github.com/NomicFoundation/hardhat/pull/6232#issuecomment-2653983599

ChristopherDedominici commented on v-next/hardhat-keystore/src/internal/keystores/encryption.ts:
```diff
@@ -0,0 +1,630 @@
+export class SecretNotFoundError extends CustomError {
+  public readonly key: string;
+
+  constructor(key: string) {
```
I agree, it is confusing. There is a work item to track it, see here: https://github.com/NomicFoundation/hardhat/pull/6232#issuecomment-2653360769

ChristopherDedominici commented on v-next/hardhat-keystore/src/internal/keystores/encryption.ts:
```diff
@@ -0,0 +1,630 @@
+/// ////////////////////////////////////////////////////////////////////////////
+
+/**
+ * Encrypts the utf-8 encoded value using the master key, and a new random iv.
```
Done here: https://github.com/NomicFoundation/hardhat/pull/6232/commits/09ff1ab810c291c6559a0579f2f75cad77422b79

ChristopherDedominici commented on v-next/hardhat-keystore/src/internal/keystores/encryption.ts:
```diff
@@ -0,0 +1,630 @@
+    throw new SecretNotFoundError(keyToRemove);
+  }
+
+  validateHmac({ masterKey, encryptedKeystore });
```
Done here: https://github.com/NomicFoundation/hardhat/pull/6232/commits/9244d97d86e2fe6848b6794a53d3f560d8f27f00

ChristopherDedominici commented on v-next/hardhat-keystore/src/internal/tasks/delete.ts:
```diff
@@ -32,7 +42,7 @@ export const remove = async (
-  const keys = await keystore.listKeys();
+  const keys = await keystore.listUnverifiedKeys();
 
   if (!keys.includes(key)) {
```
Here we check if the key we want to delete exists using the unverified method `listUnverifiedKeys`. Related to this discussion:

alcuadrado replied:
I think I'd go for asking for the password first. Failing on the side of security.

ChristopherDedominici replied:
See here: https://github.com/NomicFoundation/hardhat/pull/6232#issuecomment-2656758272

ChristopherDedominici commented:
> `UnverifiedKeyStore`

This has been addressed here: https://github.com/NomicFoundation/hardhat/pull/6232/commits/4cb0d2528e3e101d34055ecef5f068984860cb4a

The function `listKeys` has been renamed to `listUnverifiedKeys` to clearly indicate that no integrity check is performed.
The only spot where there could be an issue is here: https://github.com/NomicFoundation/hardhat/pull/6232/files#r1952835150

alcuadrado commented on v-next/hardhat-keystore/src/internal/loaders/file-manager.ts:
```diff
@@ -13,14 +14,12 @@ export class FileManagerImpl implements FileManager {
-    keystoreFile: UnencryptedKeystoreFile,
+    keystoreFile: EncryptedKeystore,
   ): Promise<void> {
     return writeJsonFile(absolutePathToFile, keystoreFile);
```
Can we write to `"${absolutePathToFile}.tmp"`, and then move that to `absolutePathToFile` once it's ready? That would minimize the chances of getting a corrupted file.

ChristopherDedominici replied:
Done here: https://github.com/NomicFoundation/hardhat/pull/6232/commits/88b9b5c12a5b73bd2f5d71f4a109096f505eaa5d

ChristopherDedominici commented:
@alcuadrado @agostbiro based on @agostbiro feedback I modified the code in [this commit ](https://github.com/NomicFoundation/hardhat/pull/6232/commits/67af98902b2313603571dafc5a221a73b4f6f1c2) in the following way:

- I added a function called `doesKeyExist` (with corresponding tests) that first checks the `HMAC integrity` and then verifies if the key exists.
- In the hook used to fetch secrets, the key existence check is now performed using the `doesKeyExist` function, ensuring the file is validated before any action is taken.
- In the `delete`, `get`, and `set` tasks, the key existence check is also done using the `doesKeyExist` function, ensuring the file is validated before performing any actions.

Please double-check that everything looks good to you.

agostbiro commented:
Thanks @ChristopherDedominici ! Keystore internals LGTM.

alcuadrado commented on v-next/hardhat-keystore/src/internal/keystores/encryption.ts:
```diff
@@ -0,0 +1,659 @@
+ * This function first calls `validateHmac` to verify the cryptographic integrity of
+ * the keystore before checking for the existence of the specified key.
+ */
+export function doesKeyExist({
```
Tiny comment: can we move this to the `Keystore primitives` section of this file? e.g. before `addSecretToKeystore`.

ChristopherDedominici replied:
Done here: https://github.com/NomicFoundation/hardhat/pull/6232/commits/c2ed9d04b3f919350945539553cbaade2fb115fd

alcuadrado reviewed (approved):
🥳
